### PR TITLE
Make updating avatar states async in battle.

### PR DESCRIPTION
**Pull request by Unengine**
### Description

SSIA

### How to test

1. Check if freezing is reduced right before the battle result is shown.

### Required Reviewers

@planetarium/9c-dev

### CODE REVIEW

File: nekoyume/Assets/_Scripts/BlockChain/ActionRenderHandler.cs [hunk 1]
```diff
@@ -551,18 +552,25 @@ private void ResponseHackAndSlash(ActionBase.ActionEvaluation<HackAndSlash> eval
...
+                                if (eval.OutputStates.TryGetAvatarStateV2(agentAddress, eval.Action.avatarAddress,
+                                    out var avatarState))
+                                {
+                                    RenderQuest(eval.Action.avatarAddress,
+                                        avatarState.questList.completedQuestIds);
+                                    _disposableForBattleEnd = null;
+                                }
+                            });
+
+                            await task.ToObservable()
```
Comment (longfin) [inline nekoyume/Assets/_Scripts/BlockChain/ActionRenderHandler.cs:571]: IMO, it seems okay without `await`. (It's all the same below.)

```suggestion
                            task.ToObservable()
```

File: nekoyume/Assets/_Scripts/BlockChain/ActionRenderHandler.cs [hunk 2]
```diff
@@ -19,6 +19,7 @@
 using Nekoyume.State.Subjects;
 using Nekoyume.UI.Module;
 using UnityEngine;
+using Cysharp.Threading.Tasks;
```
Comment (longfin) [inline nekoyume/Assets/_Scripts/BlockChain/ActionRenderHandler.cs:22]: Is it okay to add just `using`? (are we using UniTask already?)

File: nekoyume/Assets/_Scripts/BlockChain/ActionRenderHandler.cs (hunk 1 above)
Comment (Unengine) [reply nekoyume/Assets/_Scripts/BlockChain/ActionRenderHandler.cs:571]: Then, should I remove this `async` in this line?
```
.Subscribe(async _ =>
 ```

File: nekoyume/Assets/_Scripts/BlockChain/ActionRenderHandler.cs (hunk 2 above)
Comment (Unengine) [reply nekoyume/Assets/_Scripts/BlockChain/ActionRenderHandler.cs:22]: Yes, we are already using it.

File: nekoyume/Assets/_Scripts/BlockChain/ActionRenderHandler.cs (hunk 1 above)
Comment (longfin) [reply nekoyume/Assets/_Scripts/BlockChain/ActionRenderHandler.cs:571]: right.

Comment (Unengine) [reply nekoyume/Assets/_Scripts/BlockChain/ActionRenderHandler.cs]: Fixed it!

Comment (boscohyun) [reply nekoyume/Assets/_Scripts/BlockChain/ActionRenderHandler.cs]: You can use like below.
```cs
Game.Game.instance.Stage.onEnterToStageEnd
  .First()
  .Subscribe(_ => UniTask.Run(() =>
    {
      // Agent.GetState();
    })
    .ToObservable()
    .First()
    .DoOnError(Debug.LogException));
```

**Review by boscohyun — APPROVED** [review]

File: nekoyume/Assets/_Scripts/BlockChain/ActionRenderHandler.cs (hunk 1 above)
Comment (Unengine) [reply nekoyume/Assets/_Scripts/BlockChain/ActionRenderHandler.cs]: Also improved the syntax like following.

File: nekoyume/Assets/_Scripts/Game/Stage.cs [hunk 3]
```diff
@@ -533,8 +541,7 @@ private IEnumerator CoStageEnd(BattleLog log)
                 objectPool.ReleaseExcept(ReleaseWhiteList);
             }
 
-            var avatarAddress = States.Instance.CurrentAvatarState.address;
-            var avatarState = States.Instance.GetAvatarStateV2(avatarAddress);
+            var avatarAddress = avatarState.address;
```
Comment (boscohyun) [inline nekoyume/Assets/_Scripts/Game/Stage.cs:544]:
```suggestion
```

File: nekoyume/Assets/_Scripts/Game/Stage.cs [hunk 4]
```diff
@@ -470,7 +471,14 @@ private IEnumerator CoRankingBattleEnter(BattleLog log)
...
         private IEnumerator CoStageEnd(BattleLog log)
         {
+            AvatarState avatarState = null;
+            ActionRenderHandler.Instance.Pending = false;
             _onEnterToStageEnd.OnNext(this);
+            yield return ActionRenderHandler.Instance
+                .AvatarStateForBattleResultSubject
+                .First()
+                .Subscribe(value => avatarState = value);
+
```
Comment (boscohyun) [inline nekoyume/Assets/_Scripts/Game/Stage.cs:481]:
```suggestion
            AvatarState avatarState = null;
            yield return ActionRenderHandler.Instance
                .AvatarStateForBattleResultSubject
                .DoOnSubscribe(() =>
                {
                    ActionRenderHandler.Instance.Pending = false;
                    _onEnterToStageEnd.OnNext(this);
                })
                .First()
                .Subscribe(value => avatarState = value);
```

File: nekoyume/Assets/_Scripts/Game/Stage.cs [hunk 5]
```diff
@@ -622,7 +628,12 @@ private IEnumerator CoSlideBg()
 
         private IEnumerator CoRankingBattleEnd(BattleLog log, bool forceQuit = false)
         {
+            ActionRenderHandler.Instance.Pending = false;
             _onEnterToStageEnd.OnNext(this);
+            yield return ActionRenderHandler.Instance
+                .AvatarStateForBattleResultSubject
+                .First();
+
```
Comment (boscohyun) [inline nekoyume/Assets/_Scripts/Game/Stage.cs:636]:
```suggestion
            yield return ActionRenderHandler.Instance
                .AvatarStateForBattleResultSubject
                .DoOnSubscribe(() =>
                {
                    ActionRenderHandler.Instance.Pending = false;
                    _onEnterToStageEnd.OnNext(this);
                })
                .First();
```

**Review by boscohyun — CHANGES_REQUESTED** [review]

File: nekoyume/Assets/_Scripts/Game/Stage.cs (hunk 4 above)
Comment (Unengine) [reply nekoyume/Assets/_Scripts/Game/Stage.cs:481]: `ActionRenderHandler.Instance.AvatarStateForBattleResultSubject` is pre-instantiated subject. Are they also invoked on calling `DoOnSubscribe`?

File: nekoyume/Assets/_Scripts/Game/Stage.cs (hunk 5 above)
Comment (Unengine) [reply nekoyume/Assets/_Scripts/Game/Stage.cs:636]: I think this subject not subscribed here, but just waiting for message. Will those lines work?

**Review by boscohyun — APPROVED** [review]

**Comment by boscohyun** [comment]
@Unengine Please change the target branch from the `main` to the `development` branch.

**Comment by longfin** [comment]
@Unengine There is a conflicting file now.

**Pull request merged**